File: src/ObjectManager/Object.Tes/FilePacks/Records/045-IDLE.Idle Animations.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class IDLERecord : Record
    {
        public override string ToString() => $"IDLE: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL;
        public List<SCPTRecord.CTDAField> CTDAs = new List<SCPTRecord.CTDAField>(); // Conditions
        public BYTEField ANAM;
        public FMIDField<IDLERecord>[] DATAs;

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "CTDA":
                case "CTDT": CTDAs.Add(new SCPTRecord.CTDAField(r, dataSize, format)); return true;
                case "ANAM": ANAM = new BYTEField(r, dataSize); return true;
                case "DATA": DATAs = new FMIDField<IDLERecord>[dataSize >> 2]; for (var i = 0; i < DATAs.Length; i++) DATAs[i] = new FMIDField<IDLERecord>(r, 4); return true;
                    
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-IDLE.Idle Animations.swift
//
//  IDLERecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class IDLERecord: Record, IHaveEDID, IHaveMODL {
    public override var description: String { return "IDLE: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var MODL: MODLGroup? = nil
    public var CTDAs = [SCPTRecord.CTDAField]() // Conditions
    public var ANAM: BYTEField!
    public var DATAs: [FMIDField<IDLERecord>]!
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "CTDA",
             "CTDT": CTDAs.append(SCPTRecord.CTDAField(r, dataSize, format))
        case "ANAM": ANAM = r.readT(dataSize)
        case "DATA":
            DATAs = [FMIDField<IDLERecord>](); let capacity = dataSize >> 2; DATAs.reserveCapacity(capacity)
            for _ in 0..<capacity { DATAs.append(FMIDField<IDLERecord>(r, 4)) }
        default: return false
        }
        return true
    }
}
